src/backend/utils/misc/queryenvironment.c

```diff
@@ -189,23 +196,37 @@ bool has_existing_enr_relations()
 /*
  * This returns an ENR if there is a name match in the given collection.  It
  * must quietly return NULL if no match is found.
+ * 
+ * search - true if we want to include parent queryEnvs (such as in 
+ * 			babelfish_get_enr_list, object_id).
+ * 
+ * 			Other such places (like in HeapCreateWithCatalog, SPI)
+ * 			allow duplicate ENR names in the top level queryEnv, so 
+ * 			it must be false there.
  */
 EphemeralNamedRelation
-get_ENR(QueryEnvironment *queryEnv, const char *name)
+get_ENR(QueryEnvironment *queryEnv, const char *name, bool search)
```

Comment:
can we limit this special search to table variable and temp tables only? we can find that using name? also, do we have use cases where enr names could be duplicate for table variables/temp tables in same session but at different nesting level

Reply:
sure. the example I have is when running a test for BABEL-4672. without setting `search = false` as above, the following will cause a server crash

```
CREATE TRIGGER tr_emp_salary_instead_insert ON emp_salary
INSTEAD OF INSERT
AS
INSERT INTO emp_salary VALUES(2, 2000);
GO

CREATE TRIGGER tr_emp_salary_after_insert ON emp_salary
AFTER INSERT
AS
INSERT INTO emp_salary VALUES(3, 3000);
GO

INSERT INTO emp_salary VALUES (1,1000);
GO
```

Reasoning here is that it uses a default name (such as 'updated' or 'inserted') for the ENR. If it's nested, then we will run into a name collision if we do a search through all parent queryEnvs. One example of a crash caused by that is in `register_ENR` by hitting `Assert(get_ENR(queryEnv, enr->md.name, true) == NULL);`.